### PR TITLE
[Feature] Prioritized Experience Replay

### PR DESCRIPTION
**Pull request by ysqyang**
# Description

Prioritized experience replay (PER) as a subclass of AbsSampler

https://arxiv.org/pdf/1511.05952.pdf
https://github.com/rlcode/per

## Linked issue(s)/Pull request(s)

- [issue_number](issue_link)

## Type of Change

- [ ] Non-breaking bug fix
- [ ] Breaking bug fix
- [x] New feature
- [ ] Test
- [ ] Doc update
- [ ] Docker update

## Related Component

- [ ] Simulation toolkit
- [x] RL toolkit
- [ ] Distributed toolkit

## Has Been Tested

- OS:
  - [ ] Windows
  - [ ] Mac OS
  - [ ] Linux
- Python version:
  - [ ] 3.6
  - [ ] 3.7
- Key information snapshot(s):

## Needs Follow Up Actions

- [ ] New release package
- [ ] New docker image

## Checklist

- [x] Add/update the related comments
- [ ] Add/update the related tests
- [x] Add/update the related documentations
- [ ] Update the dependent downstream modules usage

### CODE REVIEW

File: maro/rl/algorithms/dqn.py [hunk 1]
```diff
@@ -19,38 +18,31 @@ class DQNConfig:
...
             Defaults to 1.0.
         double (bool): If True, the next Q values will be computed according to the double DQN algorithm,
             i.e., q_next = Q_target(s, argmax(Q_eval(s, a))). Otherwise, q_next = max(Q_target(s, a)).
             See https://arxiv.org/pdf/1509.06461.pdf for details. Defaults to False.
-        loss_cls: A string indicating a loss class provided by torch.nn or a custom loss class. If it is a string,
-            it must be a key in ``TORCH_LOSS``. Defaults to "mse".
     """
     __slots__ = [
         "reward_discount", "target_update_freq", "train_epochs", "gradient_iters", "soft_update_coefficient",
-        "double", "loss_func"
```
Comment (Jinyu-W) [inline maro/rl/algorithms/dqn.py:34]: why remove `loss_func`?

File: maro/rl/experience/sampler.py [hunk 2]
```diff
@@ -2,15 +2,131 @@
...
+            priority = self._get_priority(err)
+            tree_idx = idx + self.experience_manager.capacity - 1
+            delta = priority - self._sum_tree[tree_idx]
+            self._sum_tree[tree_idx] = priority
+            self._update(tree_idx, delta)
+
+    def get(self):
+        """Priority-based sampling."""
+        indexes, priorities = [], []
+        segment_len = self.total() / self.batch_size
```
Comment (Jinyu-W) [inline maro/rl/experience/sampler.py:89]: why split data into segments? to avoid duplicated samples?

**Review by Jinyu-W — CHANGES_REQUESTED** [review]
There are some lint issues need to fix

File: maro/rl/algorithms/dqn.py (hunk 1 above)
Comment (ysqyang) [reply maro/rl/algorithms/dqn.py:34]: The original DQN uses MSE loss (see psuedo code on Page 7 of the paper), while the prioritized experience replay version uses the vanilla TD error (see lines 11 & 13 on Page 5 of the paper) due to the need for annealing with importance sampling weights. I believe these choices are based on some mathematical guarantees (although I am not 100% sure). It is very unlikely that one would want to use a loss function other than these two.

File: maro/rl/experience/sampler.py (hunk 2 above)
Comment (ysqyang) [reply maro/rl/experience/sampler.py:89]: I believe this is called stratified sampling (see the last paragraph on Page 4 of the paper) to achieve minibatch balancing.

File: maro/rl/algorithms/dqn.py (hunk 1 above)
Comment (Jinyu-W) [reply maro/rl/algorithms/dqn.py:34]: Will it be better if we still enable this parameter and use MSE as the default loss func?

Comment (Jinyu-W) [reply maro/rl/algorithms/dqn.py:34]: Also, I remember that besides the proportional prioritized replay, there is another rank-based one. How about mentioning this in the docstring of this class?

**Comment by ysqyang** [comment]
> There are some lint issues need to fix

Fixed.

File: maro/rl/algorithms/dqn.py (hunk 1 above)
Comment (ysqyang) [reply maro/rl/algorithms/dqn.py:34]: > Will it be better if we still enable this parameter and use MSE as the default loss func?

We could. My concern is that if prioritized experience replay is used, the "loss_func" parameter would be useless.

Comment (ysqyang) [reply maro/rl/algorithms/dqn.py:34]: > Also, I remember that besides the proportional prioritized replay, there is another rank-based one. How about mentioning this in the docstring of this class?

OK. I haven't quite looked at the implementation of the rank-based algorithm, though.

**Review by Jinyu-W — APPROVED** [review]
